packages/features/insights/server/virtual-queues.ts

```diff
@@ -0,0 +1,82 @@
+import { readonlyPrisma as prisma } from "@calcom/prisma";
+import { SchedulingType } from "@calcom/prisma/enums";
+import { getSerializableForm } from "@calcom/routing-forms/lib/getSerializableForm";
+import { zodRoutes } from "@calcom/routing-forms/zod";
+
+class VirtualQueuesInsights {
+  static async getUserRelevantTeamRoutingForms({ userId }: { userId: number }) {
+    const teamRoutingForms = await prisma.app_RoutingForms_Form.findMany({
+      where: {
+        team: {
+          members: {
+            some: {
+              userId,
+            },
+          },
+        },
+      },
+      include: {
+        team: {
+          select: {
+            parentId: true,
+            parent: {
+              select: {
+                slug: true,
+              },
+            },
+            metadata: true,
+          },
+        },
+        user: {
+          select: {
+            id: true,
+            username: true,
+            movedToProfileId: true,
+          },
+        },
+      },
+    });
+
+    const serializableForms = await Promise.all(
+      teamRoutingForms.map(async (form) => {
+        const serializedForm = await getSerializableForm({ form });
+        return {
+          ...serializedForm,
+          team: form.team,
+        };
+      })
+    );
+    // find routing forms that link to RR event type
+    const formsRedirectingToWeightedRR = [];
+
+    for (const form of serializableForms) {
+      const routes = zodRoutes.parse(form.routes);
+      if (!routes) continue;
+
+      for (const route of routes) {
+        if ("action" in route) {
+          const eventType = await prisma.eventType.findFirst({
```

Comment:
We should refactor this to get this data in bulk. We shouldn’t have prisma calls inside of nested loops.

Reply:
Hey @CarinaWolli i pushed https://github.com/calcom/cal.com/pull/18260/commits/722d4c7e6c4f8621e078b3e9a8a89a3111ec2f6d to use rawSQL instead of this nasty nested loop. I've tested it and it works as expected, if you have a better approach please just revert this commit!